javalib/src/main/scala/java/net/SocketHelpers.scala

```diff
@@ -0,0 +1,185 @@
+      if (status != 0)
+        return None
+
+      var addr = stackalloc[Byte]
+      if ((!ret).ai_family == AF_INET) {
+        addr = (!ret).ai_addr.cast[Ptr[sockaddr_in]].sin_addr.cast[Ptr[Byte]]
+      } else {
+        addr = (!ret).ai_addr.cast[Ptr[sockaddr_in6]].sin6_addr.cast[Ptr[Byte]]
+      }
+      inet_ntop((!ret).ai_family, addr, ipstr, INET6_ADDRSTRLEN.toUInt)
+      freeaddrinfo(!ret)
+      Some(fromCString(ipstr))
+    }
+  }
+
+  def hostToIpArray(host: String): scala.Array[String] = {
+    Zone { implicit z =>
+      var hints = stackalloc[addrinfo]
+      var ret   = stackalloc[Ptr[addrinfo]]
+
+      string.memset(hints.cast[Ptr[Byte]], 0, sizeof[addrinfo])
+      hints.ai_family = AF_UNSPEC
+      hints.ai_socktype = SOCK_STREAM
+      hints.ai_protocol = 0
+      hints.ai_canonname = null
+      hints.ai_addr = null
+      hints.ai_next = null
+
+      val retArray = scala.collection.mutable.ArrayBuffer[String]()
+      val status   = getaddrinfo(toCString(host), null, hints, ret)
+      if (status != 0)
+        return scala.Array.empty[String]
+
+      var p = !ret
+      while (p != null) {
+        var ipstr = stackalloc[CChar](INET6_ADDRSTRLEN + 1)
+        var addr  = stackalloc[Byte]
+        if (p.ai_family == AF_INET) {
+          addr = p.ai_addr.cast[Ptr[sockaddr_in]].sin_addr.cast[Ptr[Byte]]
+        } else {
+          addr = p.ai_addr.cast[Ptr[sockaddr_in6]].sin6_addr.cast[Ptr[Byte]]
+        }
+        inet_ntop(p.ai_family, addr, ipstr, INET6_ADDRSTRLEN.toUInt)
+        retArray += fromCString(ipstr)
+        p = p.ai_next.cast[Ptr[addrinfo]]
+      }
+      freeaddrinfo(!ret)
+      retArray.toArray
+    }
+  }
+
+  def ipToHost(ip: String, isV6: Boolean): Option[String] = {
+    Zone { implicit z =>
+      var status  = 0
+      val host    = stackalloc[CChar](1024)
+      val service = stackalloc[CChar](20)
+      if (isV6) {
+        val addr6 = stackalloc[sockaddr_in6]
+        addr6.sin6_family = AF_INET6.toUShort
+        inet_pton(AF_INET6, toCString(ip), addr6.sin6_addr.cast[Ptr[Byte]])
+        status = getnameinfo(addr6.cast[Ptr[sockaddr]],
+                             sizeof[sockaddr_in6].toUInt,
+                             host,
+                             1024.toUInt,
+                             service,
+                             20.toUInt,
+                             0)
+      } else {
+        val addr4 = stackalloc[sockaddr_in]
+        addr4.sin_family = AF_INET.toUShort
+        inet_pton(AF_INET, toCString(ip), addr4.sin_addr.cast[Ptr[Byte]])
+        status = getnameinfo(addr4.cast[Ptr[sockaddr]],
+                             sizeof[sockaddr_in].toUInt,
+                             host,
+                             1024.toUInt,
+                             service,
+                             20.toUInt,
+                             0)
+      }
+
+      if (status == 0) Some(fromCString(host)) else None
+    }
+  }
+
+}
```

Comment:
This file should be in `nativelib/src/main/scala/java/net/SocketHelpers.scala`. When compiling a Scala program, it won't be on the compilation classpath, meaning that you won't be able to use it. I understand that this is an implementation detail, but since you can't use it directly, you won't be able to unit test it, and I have the feeling that this is something that could easily be broken by a refactoring, for instance.